### DISCUSSION TITLE
2.7.0 - too much? What about security?

### DISCUSSION THREAD

**Discussion by grebulon** [keepassxreboot/keepassxc#7741] [Ideas]
After looking at the 2.7.0 change log, I got worried that the project is integrating too much stuff into the code. Supporting all sorts of obscure features that may be nice but not very useful for most people. Also changing the database format which may [and according to a report here already] break compatibility with mobile apps.
Most of all I'm worried about security problems that are bound to surface when adding too much in too little time.
Personally, I'm going to switch to a different port of keepass with less features and less complexity.

**Comment by phoerious** [comment]
We are not changing the database format and many of the features KeePassXC supports actually enhance its security. Less complexity doesn't necessarily mean fewer bugs. The size of the developer team and user base are just as important.

**Comment by droidmonkey** [comment]
I have personally _removed_ thousands of lines of code when coding for 2.7.0.

**Reply by grebulon** [reply]
I have no problem with removing old code, the problem is add new stuff. I'm mainly concerned about integration with other stuff. Things that I'd look at if I wanted to find an expolit. Like: cli support, nfc, browser extensions, ssh agent, kdbx:// deep links and more. It is possible to do all of this completely secure but it's hard. And since I avoid using any of these it just seems too much for me.

**Reply by phoerious** [reply]
The CLI is a totally separate application. The rest are important features for secure password management. You can simply use PGP-encrypted text files, but that will end up being way to complicated (for the expert user, novice users will simply revert to using one password for everything instead) AND also less secure than using a full-featured password manager. The browser extension, e.g., is by design one of the most secure ways of filling passwords into websites with built-in protection against phishing sites, clickjacking etc. Oftentimes, not having these features disproportionately increases the attack surface on the side of the user themselves. When you have to decrypt a file, copy, and paste the password manually, then you trade complexity in the app for complexity in user interactions. A lot of things can go wrong there. Passwords can end up in the swap file, another app could read the clipboard, you have to decrypt the file to some unsecured location and securely delete it afterwards or find/copy/paste from a `less` window, you could accidentally paste the password into the wrong window, or at 3 a.m. in the morning, you fail to realise that paypal.com.example.org isn't actually PayPal.
